### ISSUE TITLE
[FR] add huber option for smooth_l1_loss

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#48595]
## 🚀 Feature

`SmoothL1Loss` is different from huber for just a `beta` scaling. It doesn't hurt to add an option that supports it!

cc @albanD @mruberry

**Comment by jbschlosser** [comment]
Any objection to adding a new `huber` arg (default `False`) to the existing `SmoothL1Loss`?
```python
criterion = nn.SmoothL1Loss(beta=0.5, huber=True)
```

**Comment by gchanan** [comment]
I'd start a little earlier in the discussions -- why did we add beta without huber?  (See https://github.com/pytorch/pytorch/issues/16357 for the request for beta).  Does anyone actually want `beta` and not `huber`?

Maybe @fmassa has an opinion.

**Comment by jbschlosser** [comment]
@gchanan My understanding is that Smooth L1 is mainly popular because Ross Girschick used it in the extremely influential [Fast R-CNN paper](https://arxiv.org/pdf/1504.08083.pdf). It's unclear to me why Girschick chose the variation he did rather than using the standard form of the [Huber loss](https://en.wikipedia.org/wiki/Huber_loss), but they're different by only a factor of beta, as @SsnL pointed out. To quote @fmassa from the linked issue: "The beta argument in smooth_l1_loss is the the argument which controls where the frontier between the L1 and the L2 losses are switched." So the value added with this param is to allow for tuning this frontier. Clearly, when beta=1 (as before the addition of the parameter), the two variations are equivalent.

When expanding Smooth L1 to include the beta parameter, I think a reasonable choice was made to implement the Smooth L1 variation. This is good for CV researchers, who I think would in general prefer the variation used by Girschick. For other applications, I believe the canonical Huber loss would be the go-to loss when looking for L2 behavior with less sensitivity to outliers. So one reason to provide both variations would be to match user expectations, depending on the domain they're coming from. I defer to @fmassa's opinion on whether providing both is worthwhile!

Interestingly, it looks like TensorFlow chose the Huber loss variation in [their implementation](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Huber).

**Comment by jbschlosser** [comment]
Expanding on this a bit, I believe supporting both Smooth L1 and Huber losses is useful, as I see use cases for both within Computer Vision and the broader ML / statistics community. Couple options here:

1. Add the huber flag to `SmoothL1Loss` as proposed
    Pro: Take advantage of high similarity between Smooth L1 and Huber variations - may be simpler to implement
2. New `HuberLoss` in core
    Pro: Better discoverability for users who are not familiar with the CV domain (also matches TensorFlow)

I personally like the second option, as I think `SmoothL1Loss` fits better in torchvision longer term (see https://github.com/pytorch/vision/issues/2980). Further, with separate Huber / Smooth L1 kernels for performance, the pro of Option 1 would not materialize anyway.

Any thoughts here? @albanD @fmassa

**Comment by SsnL** [comment]
I like option 2 as well!

**Comment by albanD** [comment]
Thanks for the detailed discussion @jbschlosser !
I would agree that option 2 sounds better here. It will also make sure that it is easily discoverable from our doc where the search is not the best.

**Comment by fmassa** [comment]
Hi,

I think introducing `HuberLoss` in PyTorch would be a good addition. I would even say that letting `SmoothL1Loss` delegate to `HuberLoss` and perform the extra scaling outside the kernel implementation would be fine to avoid code-duplication. I think we should also mention in the documentations that SmoothL1 is a scaled version of HuberLoss. The description in [fvcore implementation](https://github.com/facebookresearch/fvcore/blob/1dbd159daebf39dd1cb5911d5ab9b64418d0057e/fvcore/nn/smooth_l1_loss.py#L10-L36) is pretty good and we could use it to make the difference clear to the users.

The points brought by @jbschlosser are spot on. Here are a couple more points to complement his answer:

The original reason why `SmoothL1Loss` was implemented was to support Fast R-CNN (back in Lua-torch days). Fast R-CNN used only `beta=1`, and as such it was actually equivalent to Huber loss. We should have just [named it Huber loss when we added it to Lua-torch](https://github.com/torch/nn/issues/579#issuecomment-174227977) as they were both equivalent.

Later Faster R-CNN introduced the extra `beta` parameter, which didn't match the exact definition of Huber loss. IIUC, keeping the slope of the linear part of the function fixed (instead of varying it with beta as in Huber loss) allowed to more easily compare the loss values between different experiments (for example by varying beta during different runs, you wouldn't need to rebalance the other losses), and it would make it easier to validate if the non-differentiability of the L1 loss at 0 mattered or not. As @rbgirshick pointed out

> To me the really important thing is that smooth l1 loss really is *smooth l1* loss 😛. It is exactly l1 loss with the kink smoothed by a quadratic function. The degree of smoothing can be controlled by beta.
